[ENT-96] Add details to EnterpriseCustomerUser; link from Manage Learners

haikuginger opened the pull request:
This pull request enhances the EnterpriseCustomerUser admin panel to contain the following additional information:
* Username
* email address
* Added date
* Data Sharing Consent status/history

In addition, the Manage Learners template has been enhanced to link directly to each EnterpriseCustomerUser.

**JIRA tickets**: Implements ENT-96

**Dependencies**: None

**Screenshots**:

<img width="780" alt="screen shot 2016-12-15 at 4 57 02 pm" src="https://cloud.githubusercontent.com/assets/7773758/21243626/aa2a3216-c2e7-11e6-8cbb-7181099fcf05.png">


<img width="943" alt="screen shot 2016-12-15 at 4 57 29 pm" src="https://cloud.githubusercontent.com/assets/7773758/21243633/b2bbdac4-c2e7-11e6-8405-06969ff74792.png">


**Sandbox URL**: https://ent96.opencraft.hosting/ (direct IP: http://213.32.75.54; use `staff` account for superuser access)

**Merge deadline**: Jan. 9 2016

**Testing instructions**:

1. In the sandbox, go to the Manage Learner view for the existing EnterpriseCustomer
2. Observe that the username for each linked user is now a link, rather than plain text
3. Click the link, and observe that you're taken to the EnterpriseCustomerUser admin
4. Observe that user details are visible, but not editable, on the admin.
5. Observe that historical UserDataSharingConsentAudit items are listed below the user details.

**Author notes and concerns**:


1. Note that at present, a "status" item is not shown; this is because all EnterpriseCustomerUsers are linked to EnterpriseCustomers.


**Reviewers**
- [ ] @itsjeyd 
- [ ] edX reviewer[s] TBD

**Merge checklist:**
- [x] All reviewers approved
- [x] CI build is green
- [x] Version bumped
- [x] Changelog record added
- [x] Commits are (reasonably) squashed
- [x] Translations are updated
- [x] PR author is listed in AUTHORS

edx-webhook commented:
Thanks for the pull request, @haikuginger! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-enterprise&number=22"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

bradenmacdonald commented:
@haikuginger I think pending users should be clickable too, which is why the ticket is framed in terms of a profile view that includes "Status (Linked, Pending)". Obviously username and data sharing consent status would be empty for pending users.

smarnach commented:
@bradenmacdonald That was my understanding of the ticket as well.  However, it would require to introduce a new view that can show either `EnterpriseCustomerUser` instances or `PendingEnterpriseCustomerUser` instances.  The additional value provided by that view seems small in comparision to the amount of code necessary to implement this.  We should check with edX before implementing this.

smarnach commented on enterprise/admin/views.py:
```diff
@@ -62,7 +62,7 @@ def _build_context(self, request, customer_uuid):
         # TODO: pylint acts stupid - find a way around it without suppressing
         enterprise_customer = EnterpriseCustomer.objects.get(uuid=customer_uuid)  # pylint: disable=no-member
-        linked_learners = EnterpriseCustomerUser.objects.filter(enterprise_customer__uuid=enterprise_customer.uuid)
+        linked_learners = EnterpriseCustomerUser.objects.filter(enterprise_customer=enterprise_customer)
```
Just out of curiosity – does this change make any difference?  The new code is more concise, and might even result in more efficient SQL, but it should do exactly the same thing, right?

haikuginger replied:
External behavior should be identical. Should being the key word.

It's just a matter of "let's trust the Django ORM to do its job".

smarnach commented on enterprise/admin/__init__.py:
```diff
@@ -15,9 +15,9 @@
+from enterprise.models import (  # pylint:disable=no-name-in-module
     EnterpriseCustomer, EnterpriseCustomerUser, EnterpriseCustomerBrandingConfiguration,
-    EnterpriseCustomerIdentityProvider,
+    EnterpriseCustomerIdentityProvider, HistoricalUserDataSharingConsentAudit
```
The magic that automatically inserts `HistoricalUserDataSharingConsentAudit` into `models` is rather ugly.  I wasn't aware it was doing that, so this import looked a bit surprising to me.  Why do you need `HistoricalUserDataSharingConsentAudit` rather than `UserDataSharingConsentAudit` here?

haikuginger replied:
Only one `UserDataSharingConsentAudit` is attached to a given `EnterpriseCustomerUser` at a given time. We want to show a list of the changes that have been made to a given `UserDataSharingConsentAudit`, rather than just the current state, which means we need to use the historical model.

smarnach commented on enterprise/admin/__init__.py:
```diff
@@ -135,12 +165,44 @@ class EnterpriseCustomerUserAdmin(admin.ModelAdmin):
+            ) + tuple(
+                get_all_field_names(self.model)
+            )
+        return super(EnterpriseCustomerUserAdmin, self).get_readonly_fields(request, obj=obj)
```
Nit: You have the code

    super(EnterpriseCustomerUserAdmin, self).get_readonly_fields(request, obj=obj)

twice, in both branches.  The code would become more readable when written like this:

    fields = super(EnterpriseCustomerUserAdmin, self).get_readonly_fields(request, obj=obj)
    if obj:
        fields += tuple(get_all_field_names(self.model))
    return fields

haikuginger commented:
Sandbox has been updated.

e-kolpakov commented:
@haikuginger @bradenmacdonald @smarnach I would disagree with @smarnach about implementation cost - in theory it would be as simple as adding a model admin for `PendingEnterpriseCustomerUser` and resolving the link to it in "ManageLearners" view the same way it is done for linked users. I would  say it is ok to ask edX about it, but maybe it would be simpler to just implement it.

e-kolpakov commented:
:+1:

- [x] I tested this: Data sharing consent history is displayedon EnterpriseCustomerUser admin and there's a link to EnterpriseCustomerUser details page from "Manage Learners" view.
- [x] I read through the code
- [x] ~~Includes documentation~~ Too early to document.

haikuginger commented:
@e-kolpakov, you're probably right. I just added a model admin for `PendingEnterpriseCustomerUser`. I also added a list of enrolled courses at upstream request (we had been communicating back and forth about that), and also changed the link in the Manage Learners page to be on the user email, so that the links for `EnterpriseCustomerUser` and `PendingEnterpriseCustomerUser` match up in style.

You should be able to test most of that right away; I'm working on a sandbox update to fix CMS static assets so I can add a course to be listed in the admin panel. Right now, with no courses present, you should see "None" listed next to "Enrolled courses" for all of the users.

haikuginger commented:
@e-kolpakov, on the sandbox, you should now see a course ID listed for the user with username `alterego`.

e-kolpakov commented on enterprise/enrollment_api.py:
```diff
@@ -37,3 +37,11 @@ def enroll_user_in_course(user, course_details, mode):
+    """
+    Query the enrollment API to get a list of the courses a user is enrolled in.
+    """
+    client = get_enrollment_api_client()
```
@haikuginger it might be a good time to refactor this api into a class - repeating `get_enrollment_api_client` in every method is a certain code smell.

haikuginger replied:
I don't disagree.

mattdrayer replied:
+1

haikuginger commented:
@saleem-latif, @asadiqbal08, @mattdrayer, @bradenmacdonald, @kashifch for review.

haikuginger commented:
Sandbox has been updated with the latest; enrolled courses are listed by name with a link to the course info page in the LMS.

mattdrayer commented on enterprise/admin/__init__.py:
```diff
@@ -135,12 +167,89 @@ class EnterpriseCustomerUserAdmin(admin.ModelAdmin):
+
+    def username(self, ecu):
+        """
+        Return the username for the attached user.
```
Please expand the `ecu` argument name to `enterprise_customer_user` and include a description in the docstring.

mattdrayer commented on enterprise/admin/__init__.py:
```diff
@@ -135,12 +167,89 @@ class EnterpriseCustomerUserAdmin(admin.ModelAdmin):
+        """
+        return ecu.user.username
+
+    def enrolled_courses(self, ecu):
```
Please expand the `ecu` argument name to `enterprise_customer_user` and include a description in the docstring.

mattdrayer commented on enterprise/admin/__init__.py:
```diff
@@ -135,12 +167,89 @@ class EnterpriseCustomerUserAdmin(admin.ModelAdmin):
+        """
+        Return a string represented the courses enrolled in
+        """
+        courses_string = mark_safe(self.get_enrolled_course_string(ecu))
```
I always get concerned about exceptions when I see nested function calls -- do we need to protect against any errors here?

haikuginger replied:
I don't believe so. `mark_safe` just adds a `__html__()` method to the string; it doesn't perform any operations on the data.

mattdrayer commented on enterprise/admin/__init__.py:
```diff
@@ -135,12 +167,89 @@ class EnterpriseCustomerUserAdmin(admin.ModelAdmin):
+        """
+        enrolled_courses = get_enrolled_courses(user=self.username(ecu))
+        course_details = []
+        for course in enrolled_courses:
```
What if the value of enrolled_courses is `'None'`?

mattdrayer commented on enterprise/admin/__init__.py:
```diff
@@ -135,12 +167,89 @@ class EnterpriseCustomerUserAdmin(admin.ModelAdmin):
+        enrolled_courses = get_enrolled_courses(user=self.username(ecu))
+        course_details = []
+        for course in enrolled_courses:
+            course_id = course['course_details']['course_id']
```
Do we need to protect against missing fields/values here?

haikuginger replied:
The values here are pretty much fundamental to the data we're receiving; while we could suppress potential errors by doing a `.get()` call, this is one case where I'd never _expect_ to get a `KeyError`, so if there is one, as a platform administrator, I'd like to know right away, and loudly. One example of such a case would be if the API client returned a value when it should have raised its own exception instead.

I'm certainly open to making whatever change you request, though.

mattdrayer replied:
I'm okay keeping it as-is -- just wanted to point it out in case there was something more we wanted to do.

mattdrayer commented on enterprise/enrollment_api.py:
```diff
@@ -37,3 +37,25 @@ def enroll_user_in_course(user, course_details, mode):
+
+def get_full_course_details(course_id):
+    """
+    Retrieve all available details about a course.
```
Please add a description of the argument and return value to the docstring.

mattdrayer commented on enterprise/enrollment_api.py:
```diff
@@ -37,3 +37,25 @@ def enroll_user_in_course(user, course_details, mode):
+    """
+    Retrieve all available details about a course.
+    """
+    session = requests.Session()
```
This also seems like something that could be bundled up into its own class.

haikuginger commented:
@mattdrayer, I've made most of those changes, and also refactored the `enrollment_api` module out to be a bit more Pythonic.

mattdrayer commented on enterprise/edx_api.py:
```diff
@@ -0,0 +1,80 @@
+# -*- coding: utf-8 -*-
+"""
+Utilities to get details from the course catalog API.
```
This is more than just course catalog access, right?

haikuginger replied:
Ah, yes.

mattdrayer commented on enterprise/edx_api.py:
```diff
@@ -0,0 +1,80 @@
+# -*- coding: utf-8 -*-
```
So I'm slightly confused by this new module -- I do like the general idea, but I'm trying to understand if this module manages access to clients across *all* Open edX services, or if it's just for the LMS.  Currently it seems we are only concerned with LMS services, in which case we might want to rename the module to reflect the domain.

However, what might be better here is to introduce a `services` package/module which could contain all of the various adapters to communicate with remote services.

@zubair-arbi @saleem-latif @asadiqbal08 FYI

haikuginger replied:
Good call; I'll rename this to `lms_api`.

mattdrayer commented on enterprise/edx_api.py:
```diff
@@ -0,0 +1,80 @@
+
+    def enroll_user_in_course(self, username, course_id, mode):
+        """
+        Query the enrollment API to enroll the user in the course specified by course_details.
```
We're not technically "querying" in this case (queries == reports/reads), we are "commanding" (mods/writes)

mattdrayer replied:
Also I'd like to see the proper docstyle applied to these methods for arguments+returns.

haikuginger replied:
I'll make those changes; these were largely moved directly from the current position without modification.

mattdrayer commented:
Getting close!  Also, looks like a rebase+reconciliation is needed for this branch.

haikuginger commented:
@griffresch, updated sandbox here.

haikuginger commented:
Rebased from 041c27c.

haikuginger commented:
@mattdrayer, ready for another pass.

mattdrayer commented:
Looks like we need a rebase on this branch -- otherwise 👍

haikuginger commented:
Rebased from daab8c1.
